costs/es_request_constructor.go

```diff
@@ -0,0 +1,254 @@
+// date histogram aggregation on the field 'usage_start_date' with a time range of a year
+func createAggregationPerYear(_ []string) []paramAggrAndName {
+	res := make([]paramAggrAndName, 1)
+	res[0] = paramAggrAndName{
+		paramName: "year",
+		paramAggr: elastic.NewDateHistogramAggregation().
+			Field("usage_start_date").Interval("year")}
+	return res
+}
+
+// createAggregationPerTag : Creates and return a new []paramAggrAndName of size 2, which creates two
+// agregation. The first aggregation is a FilterAggregation on the field 'tag.key', and with a value of
+// the tag key passed in the parameter 'paramSplit' in the form "user:<TAG_KEY_VALUE>".
+// The second aggregation is a TermsAggregation that creates bucket aggregation on the field
+// 'tag.value'.
+// No SubAggregation is created in this fuction, as it needs to be created in the nestAggregation function
+func createAggregationPerTag(paramSplit []string) []paramAggrAndName {
+	res := make([]paramAggrAndName, 2)
+	res[0] = paramAggrAndName{
+		paramName: "tag_key",
+		paramAggr: elastic.NewFilterAggregation().
+			Filter(elastic.NewTermQuery("tag.key", fmt.Sprintf("user:%v", paramSplit[1])))}
+	res[1] = paramAggrAndName{
+		paramName: "tag_value",
+		paramAggr: elastic.NewTermsAggregation().
+			Field("tag.value").Size(aggregationMaxSize)}
+	return res
+}
+
+// createCostSumAggregation : Creates and return a new []paramAggrAndName of size 1, which creates a
+// SumAggregation on the field 'cost'
+func createCostSumAggregation(_ []string) []paramAggrAndName {
+	res := make([]paramAggrAndName, 1)
+	res[0] = paramAggrAndName{
+		paramName: "cost",
+		paramAggr: elastic.NewSumAggregation().Field("cost")}
+	return res
+}
+
+// reverseAggregationArray : reverse the paramAggrAndName array that is passed to it
+func reverseAggregationArray(aggregationArray []paramAggrAndName) []paramAggrAndName {
+	for i := len(aggregationArray)/2 - 1; i >= 0; i-- {
+		opp := len(aggregationArray) - 1 - i
+		aggregationArray[i], aggregationArray[opp] = aggregationArray[opp], aggregationArray[i]
+	}
+	return aggregationArray
+}
+
+// nestAggregation : takes a slice of paramAggrAndName type, and will nest the different aggregations.
+// Aggregations are nested by creating a chain of SubAggregation
+// A type switch is required for simulating downcasting from the interface elastic.Aggregation.
+// Current types on the type switch are TermsAggregation, FilterAggregation, SumAggregation and
+// DateHistogramAggregation.
+// If a new function creating a type that is not listed here is added to the paramNameToFuncPtr map
+// it should be added to the type switch, or the function will create bugged SubAggregations
+func nestAggregation(allAggrSlice []paramAggrAndName) elastic.Aggregation {
+	allAggrSlice = reverseAggregationArray(allAggrSlice)
+	aggrToNest := allAggrSlice[0]
+	for _, baseAggr := range allAggrSlice[1:] {
+		switch assertedBaseAggr := baseAggr.paramAggr.(type) {
```

Comment:
You should use interfaces instead of a switch to cast.

Reply:
Because of the way the `SubAggregation` function is prototyped, it is not possible to add it to a new interface to be able to free ourselves from a type switch